src/vm/database/marf.rs

```diff
@@ -709,3 +660,233 @@ impl ClarityBackingStore for NullBackingStore {
+        //    the side_store shouldn't commit data for blocks that won't be
+        //    included in the processed chainstate (like a block constructed during mining)
+        //    _if_ for some reason, we do want to be able to access that mined chain state in the future,
+        //    we should probably commit the data to a different table which does not have uniqueness constraints.
+        SqliteConnection::drop_metadata(self.marf.sqlite_tx(), &self.chain_tip);
+        let _ = self.marf.commit_mined(will_move_to).map_err(|e| {
+            error!(
+                "Failed to commit to mined MARF block {}: {:?}",
+                &will_move_to, &e
+            );
+            panic!();
+        });
+    }
+
+    // This function *should not* be called by
+    //   a smart-contract, rather it should only be used by the VM
+    pub fn get_root_hash(&mut self) -> TrieHash {
+        self.marf
+            .get_root_hash_at(&self.chain_tip)
+            .expect("FATAL: Failed to read MARF root hash")
+    }
+}
+
+impl<'a> ClarityBackingStore for WritableMarfStore<'a> {
+    fn set_block_hash(&mut self, bhh: StacksBlockId) -> Result<StacksBlockId> {
+        self.marf
+            .check_ancestor_block_hash(&bhh)
+            .map_err(|e| match e {
+                MarfError::NotFoundError => {
+                    test_debug!("No such block {:?} (NotFoundError)", &bhh);
+                    RuntimeErrorType::UnknownBlockHeaderHash(BlockHeaderHash(bhh.0))
+                }
+                MarfError::NonMatchingForks(_bh1, _bh2) => {
+                    test_debug!(
+                        "No such block {:?} (NonMatchingForks({}, {}))",
+                        &bhh,
+                        BlockHeaderHash(_bh1),
+                        BlockHeaderHash(_bh2)
+                    );
+                    RuntimeErrorType::UnknownBlockHeaderHash(BlockHeaderHash(bhh.0))
+                }
+                _ => panic!("ERROR: Unexpected MARF failure: {}", e),
+            })?;
+
+        let result = Ok(self.chain_tip);
+        self.chain_tip = bhh;
+
+        result
+    }
+
+    fn get(&mut self, key: &str) -> Option<String> {
+        trace!("MarfedKV get: {:?} tip={}", key, &self.chain_tip);
+        self.marf
+            .get(&self.chain_tip, key)
+            .or_else(|e| match e {
+                MarfError::NotFoundError => {
+                    trace!(
+                        "MarfedKV get {:?} off of {:?}: not found",
+                        key,
+                        &self.chain_tip
+                    );
+                    Ok(None)
+                }
+                _ => Err(e),
+            })
+            .expect("ERROR: Unexpected MARF Failure on GET")
+            .map(|marf_value| {
+                let side_key = marf_value.to_hex();
+                trace!("MarfedKV get side-key for {:?}: {:?}", key, &side_key);
+                SqliteConnection::get(self.marf.sqlite_tx(), &side_key).expect(&format!(
```

Comment:
Just so I understand -- this is a DB transaction instead of a normal connection in order to ensure that a concurrent read and write (e.g. from two separate threads in the node) are always serialized?

Reply:
Or, am I misreading this, and both the `marf.get()` and the `SqliteConnection::get()` happen within the same transaction already?

Reply:
> both the marf.get() and the SqliteConnection::get() happen within the same transaction already?

Yep, that's correct --- `marf.sqlite_tx()` here is just returning a pointer to the currently open transaction. The `WritableMarf` struct here shares the same lifetime as the underlying SQLite transaction. Every read and write on the same instance of that struct shares a single transaction. That transaction is used not only for serialization, but atomization of the side store operations with the marf operations (it ensures that when the trie commits, the side store commits as well).